GCEWindowsAgent/oslogin.go

```diff
@@ -0,0 +1,339 @@
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+var (
+	googleComment    = "# Added by Google Compute Engine OS Login."
+	googleBlockStart = "#### Google OS Login control. Do not edit this section. ####"
+	googleBlockEnd   = "#### End Google OS Login control section. ####"
+)
+
+type osloginMgr struct{}
+
+func (a *osloginMgr) diff() bool {
+	// True on first run.
+	return oldMetadata.Project.ProjectID == "" ||
+		// True if any value has changed.
+		(oldMetadata.Instance.Attributes.EnableOSLogin != newMetadata.Instance.Attributes.EnableOSLogin) ||
+		(oldMetadata.Instance.Attributes.TwoFactor != newMetadata.Instance.Attributes.TwoFactor) ||
+		(oldMetadata.Project.Attributes.EnableOSLogin != newMetadata.Project.Attributes.EnableOSLogin) ||
+		(oldMetadata.Project.Attributes.TwoFactor != newMetadata.Project.Attributes.TwoFactor)
+}
+
+func (a *osloginMgr) timeout() bool {
+	return false
+}
+
+func (a *osloginMgr) disabled(os string) bool {
+	return os == "windows"
+}
+
+func (a *osloginMgr) set() error {
+	oldenable := oldMetadata.Instance.Attributes.EnableOSLogin || oldMetadata.Project.Attributes.EnableOSLogin
+	enable := newMetadata.Instance.Attributes.EnableOSLogin || newMetadata.Project.Attributes.EnableOSLogin
+	twofactor := newMetadata.Instance.Attributes.TwoFactor || newMetadata.Project.Attributes.TwoFactor
+
+	if enable && !oldenable {
+		logger.Infof("Enabling OS Login")
+		newMetadata.Instance.Attributes.SSHKeys = nil
+		newMetadata.Project.Attributes.SSHKeys = nil
+		(&linuxAccountsMgr{}).set()
+	}
+
+	if err := updateSSHConfig(enable, twofactor); err != nil {
+		logger.Errorf("error updating SSH config: %v\n", err)
+	}
+
+	if err := updateNSSwitchConfig(enable); err != nil {
+		logger.Errorf("error updating NSS config: %v\n", err)
+	}
+
+	if err := updatePAMConfig(enable, twofactor); err != nil {
+		logger.Errorf("error updating PAM config: %v\n", err)
+	}
+
+	if err := createOSLoginDirs(); err != nil {
+		logger.Errorf("error creating OS Login directory: %v\n", err)
+	}
+
+	if err := createOSLoginSudoersFile(); err != nil {
+		logger.Errorf("error creating OS Login sudoers file: %v\n", err)
+	}
+
+	// Services which need to be restarted primarily due to caching issues.
+	for _, svc := range []string{"ssh", "sshd", "nscd", "unscd", "systemd-logind", "cron", "crond"} {
+		if err := restartService(svc); err != nil {
+			logger.Errorf("error restarting service: %v\n", err)
+		}
+	}
+
+	if enable {
+		if err := exec.Command("google_oslogin_nss_cache").Run(); err != nil {
+			logger.Errorf("error updating NSS cache: %v\n", err)
+		}
+	}
+	os.Exit(1)
+
+	return nil
+}
+
+func filterGoogleLines(contents string) []string {
+	var isgoogle, isgoogleblock bool
+	var filtered []string
+	for _, line := range strings.Split(contents, "\n") {
+		if strings.Contains(line, googleComment) {
```

Comment:
A switch statement with a default of filtered = append(filtered, line) would be easier to read and drop the need for all the continues